### ISSUE TITLE
Smart table sorting not working for ajax loaded data

### ISSUE THREAD

**Issue by cloud9amit** [lorenzofox3/Smart-Table#156]
Hi,

I have created a plunker link: http://plnkr.co/edit/PBX00TGC5E1UEWzRAmPU?p=preview

to show that smart table data get's reset to blank once we click header for sorting. This issue is particularly seen when data is loaded using ajax.

Any quick resolution would be really appreciated.

Thanks
Amit

**Comment by Keshavmca12** [comment]
Hi
I am using angular smart table and used st-table="displayedCollection" st-safe-src="rowCollection" both but st-sort not working in my case.Will you please suggest some fix of it.

**Comment by calvinKG** [comment]
Hi, 

I am having the same problem as Keshavmca12

**Comment by starbit** [comment]
Same problem as the two guys above.

**Comment by agorm** [comment]
Me too.

**Comment by stoplion** [comment]
+1

**Comment by proteo5** [comment]
In my case it wasn't very clear on the documentation that you have to check that  the:
<tr ng-repeat="row in rowCollection">
have to chenge to 
<tr ng-repeat="row in displayedCollection">
when using  st-safe-src.

**Comment by DeBraid** [comment]
Wrestling with this as we speak, I am really saddened that there is so much config required on an otherwise excellent table package.  

This is a really core feature to any data table, how is this not enabled out of the box?

**Comment by lorenzofox3** [comment]
I don't think there is much configuration needed 
as discussed in many issues

1. st-table is just a way to declare a variable name to expose it to the view
2. st-safe-src is the only thing you have to set in the controller if you want smart table to keep tracks of external changes, that's it 

http://plnkr.co/edit/UiPpWX9nUzRXopayLOxY?p=preview

Obviously the doc could be improved though, any suggestion ?

**Comment by DeBraid** [comment]
Thanks for the plunker.  I have wired everything up as shown, but the result is randomly sorted data (the rows will change/move on click events, but the order is not correct). 

As for semantics of `st-safe-src`, it might be better to have something like `st-table-data-from-XHR` and `st-table-placeholder`.  While these are much more verbose they are more semantic and would help with picking up the nomenclature.

**Comment by rodrigoroma** [comment]
+1 for proteo5 comment. And thank you for this comment!

**Comment by rodrigoroma** [comment]
I think the ideal solution to handle this is not to pass 2 collections in the cofiguration but to create a new attribute that is a flag to indicate if the user plans to update the original collection or not. The ST should handle this by itself and probably with no loss in performance. For example:

`<table st-table="rowCollection" st-watch-changes="true">`

Well, I sent a pull request that can improve the doc about this attribute.

**Comment by viniciuscb** [comment]
I had the same problem too, it was not obvious in the documentation, however solved when following the instructions in the pluckr.

**Comment by rodrigoroma** [comment]
I did a pull request improving this part of the documentation but it's still in the "opened" status.

**Comment by georules** [comment]
I agree that this may just be an issue with documentation/clarification of how to use this feature.  When I set st-table to the same variable as st-safe-src, my sorting on async data started working.

**Comment by khullah** [comment]
Also, it wasn't clear for me in the docs that along with st-safe-src you actually have to set the array in the controller (thought it was automatic, creating a copy on the fly).

By the way, couldn't this be done? It would be a lot easier

**Comment by pgksunilkumar** [comment]
if you are sorting the number check the highest number length and add into the rest of the numbers add
ex: greater than number digits length of 4  "9999" and lowest number of digits length of 2  "0099"
```
String.prototype.lpad = function(padString, length) {
    var str = this;
    while (str.length < length)
        str = padString + str;
    return str;
}
var str = "5";
alert(str.lpad("0", 4)); //result "0005"
```
Check the Example [http://plnkr.co/edit/vjcLDWYPoh5i8SQRK0XR?p=preview]()
